Userland/Libraries/LibWeb/CSS/CountersSet.cpp

```diff
@@ -0,0 +1,103 @@
+ */
+
+#include "CountersSet.h"
+#include <LibWeb/DOM/Element.h>
+#include <LibWeb/DOM/Node.h>
+
+namespace Web::CSS {
+
+// https://drafts.csswg.org/css-lists-3/#instantiate-counter
+Counter& CountersSet::instantiate_a_counter(FlyString name, i32 originating_element_id, bool reversed, Optional<CounterValue> value)
+{
+    // 1. Let counters be element’s CSS counters set.
+    auto* element = DOM::Node::from_unique_id(originating_element_id);
+
+    // 2. Let innermost counter be the last counter in counters with the name name.
+    //    If innermost counter’s originating element is element or a previous sibling of element,
+    //    remove innermost counter from counters.
+    auto innermost_counter = last_counter_with_name(name);
+    if (innermost_counter.has_value()) {
+        auto* originating_node = DOM::Node::from_unique_id(innermost_counter->originating_element_id);
+        VERIFY(originating_node);
+        auto& innermost_element = verify_cast<DOM::Element>(*originating_node);
+
+        if (&innermost_element == element
+            || (innermost_element.parent() == element->parent() && innermost_element.is_before(*element))) {
+
+            m_counters.remove_first_matching([&innermost_counter](auto& it) {
+                return it.name == innermost_counter->name
+                    && it.originating_element_id == innermost_counter->originating_element_id;
+            });
+        }
+    }
+
+    // 3. Append a new counter to counters with name name, originating element element,
+    //    reversed being reversed, and initial value value (if given)
+    m_counters.append({
+        .name = move(name),
+        .originating_element_id = originating_element_id,
+        .reversed = reversed,
+        .value = value,
+    });
+
+    return m_counters.last();
+}
+
+// https://drafts.csswg.org/css-lists-3/#propdef-counter-set
+void CountersSet::set_a_counter(FlyString name, i32 originating_element_id, CounterValue value)
+{
+    if (auto existing_counter = last_counter_with_name(name); existing_counter.has_value()) {
+        existing_counter->value = value;
+        return;
+    }
+
+    // If there is not currently a counter of the given name on the element, the element instantiates
+    // a new counter of the given name with a starting value of 0 before setting or incrementing its value.
+    // https://drafts.csswg.org/css-lists-3/#valdef-counter-set-counter-name-integer
+    auto& counter = instantiate_a_counter(name, originating_element_id, false, 0);
+    counter.value = value;
+}
+
+// https://drafts.csswg.org/css-lists-3/#propdef-counter-increment
+void CountersSet::increment_a_counter(FlyString name, i32 originating_element_id, CounterValue amount)
+{
+    if (auto existing_counter = last_counter_with_name(name); existing_counter.has_value()) {
+        // FIXME: How should we handle existing counters with no value? Can that happen?
+        VERIFY(existing_counter->value.has_value());
```

Comment:
I ran this PR against the WPT tests in `css/css-lists` and this condition got hit in 14 tests. I've not investigated why this happens yet.

Here's a test that hits this condition: https://wpt.live/css/css-lists/counter-reset-reversed-not-list-item.html

I ran the tests with this command: `Tests/LibWeb/WPT/wpt/wpt run --webdriver-binary "${LADYBIRD_SOURCE_DIR}/Build/ladybird/bin/WebDriver" --webdriver-arg="--enable-qt-networking" --log-mach output.log --no-restart-on-unexpected --processes $(nproc) ladybird css/css-lists`

Reply:
:thinking: You know what, I never added proper support for `reversed()` which means those counters always have no value here. Thanks, I need to start getting in the habit of running WPT tests!